src/lib/ThumbnailsSidebar.js

```diff
@@ -218,30 +220,77 @@ class ThumbnailsSidebar {
             return Promise.resolve(this.thumbnailImageCache[itemIndex]);
         }
 
+        return this.getThumbnailDataURL(itemIndex + 1)
+            .then(this.createImageEl)
+            .then((imageEl) => {
+                // Cache this image element for future use
+                this.thumbnailImageCache[itemIndex] = imageEl;
+
+                return imageEl;
+            });
+    }
+
+    /**
+     * Given a page number, generates the image data URL for the image of the page
+     * @param {number} pageNum  - The page number of the document
+     * @return {string} The data URL of the page image
+     */
+    getThumbnailDataURL(pageNum) {
         const canvas = document.createElement('canvas');
 
         return this.pdfViewer.pdfDocument
-            .getPage(itemIndex + 1)
+            .getPage(pageNum)
             .then((page) => {
-                const viewport = page.getViewport(1);
-                canvas.width = THUMBNAIL_WIDTH_MAX;
-                canvas.height = THUMBNAIL_WIDTH_MAX / this.pageRatio;
-                const scale = THUMBNAIL_WIDTH_MAX / viewport.width;
+                const { width, height } = page.getViewport(1);
+                // Get the current page w:h ratio in case it differs from the first page
+                const curPageRatio = width / height;
+
+                let canvasWidth;
+                // Handle the case where the current page's w:h ratio is less than the
+                // `pageRatio` which means that this page is probably more portrait than
+                // landscape
+                if (curPageRatio < this.pageRatio) {
+                    // Set the canvas height to that of the thumbnail max height
+                    canvas.height = THUMBNAIL_WIDTH_MAX / this.pageRatio;
+                    // Find the canvas width based on the curent page ratio
+                    canvasWidth = canvas.height * curPageRatio;
+                    canvas.width = canvasWidth;
+                } else {
+                    // In case the current page ratio is same as the first page
+                    // or in case it's larger (which means that it's wider), keep
+                    // the width at the max thumbnail width
+                    canvasWidth = THUMBNAIL_WIDTH_MAX;
+                    canvas.width = canvasWidth;
+                    // Find the height based on the current page ratio
+                    canvas.height = THUMBNAIL_WIDTH_MAX / curPageRatio;
+                }
+
+                // The amount for which to scale down the current page
+                const scale = canvasWidth / width;
```

Comment:
Are we caching `canvasWidth` because calling `canvas.width` has a performance penalty? Or can we simplify and just pull the value directly?